AccountMasterKey.js

```diff
@@ -0,0 +1,165 @@
+    this.accountId = accountId;
+    this.signer = signer;
+    this.kmsService = kmsService;
+    this.kmsPlugin = kmsPlugin;
+  }
+
+  /**
+   * Generates a key. The key can be a key encryption key (KEK) or an HMAC
+   * key. It can be generated using a FIPS-compliant algorithm or the latest
+   * recommended algorithm.
+   *
+   * @param {String} type the type of key to create (`hmac` or `kek`).
+   * @param {String} version `fips` to use FIPS-compliant ciphers,
+   *   `recommended` to use the latest recommended ciphers.
+   *
+   * @return {Promise<Object>} resolves to a Kek or Hmac instance.
+   */
+  async generateKey({type, version = 'recommended'}) {
+    _assertVersion(version);
+
+    // for the time being, fips and recommended are the same; there is no
+    // other standardized key wrapping algorithm
+    let Class;
+    if(type === 'hmac') {
+      type = 'HS256';
+      Class = Hmac;
+    } else if(type === 'kek') {
+      type = 'AES-KW';
+      Class = Kek;
+    } else {
+      throw new Error(`Unknown key type "${type}".`);
+    }
+
+    const {kmsService, kmsPlugin, kmsPlugin: plugin, signer} = this;
+    const id = await kmsService.generateKey({plugin, type, signer});
+    return new Class({id, kmsService, kmsPlugin, signer});
+  }
+
+  /**
+   * Gets a KEK API for wrapping and unwrapping cryptographic keys. The key
+   * ID is presumed to be scoped to the KMS service and plugin assigned
+   * to this account master key instance.
+   *
+   * @param {String} id the ID of the KEK.
+   *
+   * @return {Promise<Kek>} the new Kek instance.
+   */
+  async getKek({id}) {
+    const {kmsService, kmsPlugin, signer} = this;
+    // FIXME: call kmsService.exists()? ...get algorithm?
+    return new Kek({id, kmsService, kmsPlugin, signer});
+  }
+
+  /**
+   * Gets an HMAC API for signing and verifying cryptographic keys. The key
+   * ID is presumed to be scoped to the KMS service and plugin assigned
+   * to this account master key instance.
+   *
+   * @param {String} id the ID of the HMAC key.
+   *
+   * @return {Promise<Hmac>} the new Hmac instance.
+   */
+  async getHmac({id}) {
+    const {kmsService, kmsPlugin, signer} = this;
+    // FIXME: call kmsService.exists()? ...get algorithm?
+    return new Hmac({id, kmsService, kmsPlugin, signer});
+  }
+
+  /**
+   * Generates a the master key from a secret.
+   *
+   * @param {String|Uint8Array} secret the secret to use (e.g. a bcrypt hash).
+   * @param {String} accountId the ID of the account associated with this
+   *   master key.
+   * @param {KmsService} kmsService the kmsService to use to perform key
+   *   operations.
+   * @param {String} kmsPlugin the ID of the KMS plugin to use.
+   *
+   * @return {Promise<AccountMasterKey>} the new AccountMasterKey instance.
+   */
+  static async fromSecret({secret, accountId, kmsService, kmsPlugin}) {
+    if(typeof secret === 'string') {
+      secret = _strToUint8Array(secret);
```

Comment:
I really don't like this sort of mutation.  Seems particularly bad in a security lib.

Reply:
Nevermind, not an issue.

Reply:
This isn't a mutation; it's using a shadowed variable.